commit 47416585a6af7b4ade018ba5c6cb0721b95d8d12
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2019-05-22T14:38:54+00:00 (5 months ago)
    
    TRITON-1692 linux-prepare-image should be able to set root password hash
    Reviewed by: Pedro Palazón Candel <pedro@joyent.com>
    Approved by: Pedro Palazón Candel <pedro@joyent.com>

diff --git a/tools/prepare-image/linux-prepare-image b/tools/prepare-image/linux-prepare-image
index 597134c..6e355ad 100755
--- a/tools/prepare-image/linux-prepare-image
+++ b/tools/prepare-image/linux-prepare-image
@@ -61,6 +61,13 @@ function errexit
 
 #---- support routines
 
+# A wrapper around mdata-get to get values from the prepare-image namespace.
+function mdata_get_pi() {
+	typeset key=$1
+
+	mdata-get "prepare-image:$key" 2>/dev/null || true
+}
+
 function cleanup_logs() {
     # This ensures we don't delete the following
     # /var/log/wtmp - syslog won't start if this is missin
@@ -91,8 +98,9 @@ function cleanup_root() {
     history -c
     history -w || true
 
-    # Removing password for root
-    passwd -d root
+    # Remove or set password for root.  If root-password-hash is not set, this
+    # is equivalent to "passwd -d root".
+    usermod -p "$(mdata_get_pi root-password-hash)" root
 
     # Clean up history, ssh keys, etc.
     cleanup_homedir /root
@@ -194,7 +202,7 @@ function prepare_ubuntu() {
     # Clean up package cache
     apt-get -y clean
 
-    local fixconsole=$(mdata-get prepare-image:fix-ubuntu-console 2>/dev/null)
+    local fixconsole=$(mdata_get_pi fix-ubuntu-console)
 
     if [[ $fixconsole == true ]]; then
         cat >/etc/default/grub.d/51-smartos-serial-console.cfg <<EOF
